runtime/core/decoder/onnx_asr_model.cc

```diff
@@ -0,0 +1,362 @@
+                                       encoder_output_size_ / head_ * 2};
+    att_cache_ort_ = Ort::Value::CreateTensor<float>(
+        memory_info_, att_cache_.data(), att_cache_.size(), att_cache_shape, 4);
+
+    att_mask_.resize(0);
+    input_names_.erase(input_names_.begin() + 2);
+    input_names_.pop_back();
+  }
+
+  cnn_cache_.resize(num_blocks_ * encoder_output_size_ *
+                    (cnn_module_kernel_ - 1));
+  const int64_t cnn_cache_shape[] = {num_blocks_, 1, encoder_output_size_,
+                                     cnn_module_kernel_ - 1};
+
+  cnn_cache_ort_ = Ort::Value::CreateTensor<float>(
+      memory_info_, cnn_cache_.data(), cnn_cache_.size(), cnn_cache_shape, 4);
+
+  encoder_outs_.clear();
+}
+
+void OnnxAsrModel::ForwardEncoderFunc(
+    const std::vector<std::vector<float>>& chunk_feats,
+    std::vector<std::vector<float>>* out_prob) {
+  int num_frames = cached_feature_.size() + chunk_feats.size();
+  const int feature_dim = chunk_feats[0].size();
+  int64_t requried_cache_size = chunk_size_ * num_left_chunks_;
+  std::vector<float> model_input;
+  for (size_t i = 0; i < cached_feature_.size(); ++i) {
+    for (int j = 0; j < feature_dim; j++) {
+      model_input.emplace_back(cached_feature_[i][j]);
+    }
+  }
+  for (size_t i = 0; i < chunk_feats.size(); ++i) {
+    for (int j = 0; j < feature_dim; j++) {
+      model_input.emplace_back(chunk_feats[i][j]);
+    }
+  }
+  int num_fill =
+      att_mask_.size() + 1 + right_context_ - subsampling_rate_ - num_frames;
+  if (num_fill > 0) {
+    num_frames += num_fill;
+    for (size_t i = 0; i < num_fill; ++i) {
+      for (int j = 0; j < feature_dim; j++) {
+        model_input.emplace_back(0);
+      }
+    }
+  }
+  const int64_t input_shape[3] = {1, num_frames, feature_dim};
+  Ort::Value input_tensor = Ort::Value::CreateTensor<float>(
+      memory_info_, model_input.data(), model_input.size(), input_shape, 3);
+  Ort::Value offset_tensor = Ort::Value::CreateTensor<int64_t>(
+      memory_info_, &offset_, 1, std::vector<int64_t>{}.data(), 0);
+
+  Ort::Value requried_cache_size_tensor = Ort::Value::CreateTensor<int64_t>(
+      memory_info_, &requried_cache_size, 1, std::vector<int64_t>{}.data(), 0);
+
+  std::vector<Ort::Value> tensors;
+  tensors.emplace_back(std::move(input_tensor));
+  tensors.emplace_back(std::move(offset_tensor));
+  if (att_mask_.empty()) {
+    tensors.emplace_back(std::move(att_cache_ort_));
+    tensors.emplace_back(std::move(cnn_cache_ort_));
+  } else {
+    const int64_t att_mask_shape[] = {1, 1, att_mask_.size()};
+    att_mask_ort_ = Ort::Value::CreateTensor<bool>(
+        memory_info_, reinterpret_cast<bool*>(att_mask_.data()),
+        att_mask_.size(), att_mask_shape, 3);
+    tensors.emplace_back(std::move(requried_cache_size_tensor));
+    tensors.emplace_back(std::move(att_cache_ort_));
+    tensors.emplace_back(std::move(cnn_cache_ort_));
+    tensors.emplace_back(std::move(att_mask_ort_));
+  }
```

Comment:
emplace_back ort_tensors according to input node names, in this way there is no need to manually control what should be emplaced:

```cpp
std::vector<Ort::Value> inputs;
  for (auto name : encoder_in_names_) {
    if (!strcmp(name, "chunk")) {
      inputs.emplace_back(std::move(feats_ort));
    } else if (!strcmp(name, "offset")) {
      inputs.emplace_back(std::move(offset_ort));
    } else if (!strcmp(name, "required_cache_size")) {
      inputs.emplace_back(std::move(required_cache_size_ort));
    } else if (!strcmp(name, "att_cache")) {
      inputs.emplace_back(std::move(att_cache_ort_));
    } else if (!strcmp(name, "cnn_cache")) {
      inputs.emplace_back(std::move(cnn_cache_ort_));
    } else if (!strcmp(name, "att_mask")) {
      inputs.emplace_back(std::move(att_mask_ort));
    }
  }
```

Reply:
since we have so many decoding cases, i.e.,  streaming model with 16/-1 16/4/ -1/-1 16/0, non-streaming model with -1/-1, different case has different in/out nodes, it is hard to manually decide which node(tensor) should be emplaced.